src/commands/classified/external.rs

```diff
@@ -49,157 +48,245 @@ impl Decoder for LinesCodec {
     }
 }
 
-#[derive(Debug)]
-pub(crate) enum StreamNext {
-    Last,
-    External,
-    Internal,
-}
-
 pub(crate) async fn run_external_command(
     command: ExternalCommand,
     context: &mut Context,
-    input: ClassifiedInputStream,
-    stream_next: StreamNext,
-) -> Result<ClassifiedInputStream, ShellError> {
-    let stdin = input.stdin;
-    let inputs: Vec<Value> = input.objects.into_vec().await;
-
+    input: InputStream,
+    is_last: bool,
+) -> Result<InputStream, ShellError> {
     trace!(target: "nu::run::external", "-> {}", command.name);
-    trace!(target: "nu::run::external", "inputs = {:?}", inputs);
 
-    let mut arg_string = command.name.to_owned();
-    for arg in command.args.iter() {
-        arg_string.push_str(&arg);
+    let has_it_arg = command.args.iter().any(|arg| arg.contains("$it"));
+    if has_it_arg {
+        run_with_iterator_arg(command, context, input, is_last).await
+    } else {
+        run_with_stdin(command, context, input, is_last).await
     }
+}
 
-    let home_dir = dirs::home_dir();
+async fn run_with_iterator_arg(
```

Comment:
There's some overlap in here with `run_with_stdin`, but given how differently these two work I wanted to first split them up, get everything working, and then see what common shapes exist.